kernel/src/loader/mod.rs

```diff
@@ -66,23 +66,26 @@ impl fmt::Display for Error {
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-/// Loads a kernel from a vmlinux elf image to a slice
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+/// Loads a kernel from a vmlinux elf image using a given callback.
 ///
 /// # Arguments
 ///
-/// * `guest_mem` - The guest memory region the kernel is written to.
 /// * `kernel_image` - Input vmlinux image.
 /// * `start_address` - For x86_64, this is the start of the high memory. Kernel should reside above it.
+/// * `write_to_memory` - Closure to write the contents of the image.
+///                      The closure is called with the following arguments:
+///                      dst-offset-in-guest-mem, src, size
 ///
 /// Returns the entry address of the kernel.
-#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-pub fn load_kernel<F>(
-    guest_mem: &GuestMemory,
-    kernel_image: &mut F,
+pub fn load_kernel<R, F>(
```

Comment:
Finally had a chance to take a better look at this and it seems there's no reason really to prevent the `loader` -> `memory_model` dependency. What the kernel loader is doing (particularly when loading elf images) is to write byte sequences (the elf program headers) at different guest memory addresses, and `GuestMemory` is the right abstraction to apply these operations to (in the `vm-memory` crate from `rust-vmm`, `GuestMemory` becomes a trait which can be implemented by multiple backends for different purposes). It looks like using closures adds some extra complexity and obfuscation to the code for no real benefit in this particular case, so we're not really improving the status quo with these changes. We can take the alternative approaches to the discussions around `linux-loader` on `rust-vmm` once the initial PR lands.